### DISCUSSION TITLE
What is "User authorization callback URL" for?

### DISCUSSION THREAD

**Discussion by unformatt** [community/community#24665] [Apps, API and Webhooks]
<p>I’ve set “User authorization callback URL”  in my “Github App” settings and when a user installed my app, they are not taken to that URL.</p>
<ul>
<li>
<p>User goes to <a href="https://github.com/apps/board-genius-dev/installations/new" rel="nofollow noopener">https://github.com/apps/[my app name]/installations/new</a></p>
</li>
<li>
<p>User select which repos my app has access to, clicks “Install”</p>
</li>
<li>
<p>User is then taken to <a href="https://github.com/settings/installations/935802" rel="nofollow noopener">https://github.com/settings/installations/xxxx</a></p>
</li>
</ul>
<p>I expected the user to be taken to the URL I’ve entered for “User authorization callback URL” not taken to the github installations page.</p>
<p>I tried to set the “Setup URL (optional)” and this gives me the desired behavior.</p>
<p>What is “User authorization callback URL” for?</p>

**Comment by AndreaGriffiths11** [comment]
<p>Hi <a class="mention" href="/unformatt">@unformatt</a>,</p>
<p>Thanks for being here! Per the docs did you set up a route to specify what the <a href="https://developer.github.com/v3/guides/basics-of-authentication/#providing-a-callback" rel="nofollow noopener">callback</a> should do?</p>

**Comment by unformatt** [comment]
<p>Hi AndreaG,</p>
<p>Yes, my server is ready to accept the callback URL. However, Github is not redirecting the user to the callback URL. It’s not a matter of 404 as mentioned in your link. Github is redirecting back to Github, not my site.</p>
<p>Thanks</p>
<p>Matt</p>

**Comment by jonathan-marcus** [comment]
<p>Hi <a class="mention" href="/unformatt">@unformatt</a>,</p>
<p>Did you ever solve this? I have the exact same problem - the “User authorization callback URL” appears to be a no-op. The user is never brought there, and no messages are sent to that URL, it just redirects to <a href="https://github.com/settings/installations/xxxx" rel="nofollow noopener">https://github.com/settings/installations/xxxx</a>.</p>

**Comment by unformatt** [comment]
<p><a class="mention" href="/jonathan-marcus">@jonathan-marcus</a> - I can’t remember how I got around it. I think when the user first installs the apps, it works (i.e. redirects) but after that, if they just updated settings, it does not redirect. I never found a way around it. Very bad UX for my users.</p>

**Comment by jonathan-marcus** [comment]
<p>Yeah, it’s not good. I don’t remember all the different combinations I’ve tried, but I currently have “Request user authorization (OAuth) during installation” checked and it does send me to this URL. I tried having the “Setup URL” specified also, but I was never able to get it really correct. I am okay right now, I do get a redirect after the GitHub app is installed, but it feels very fragile.</p>

**Comment by jamesmartin** [comment, marked as answer]
<p>:wave: <a class="mention" href="/unformatt">@unformatt</a>!</p>
<p>There are two types of API authorization for GitHub Apps:</p>
<ol>
<li>Server-to-server, where the App acts as itself and authenticates using an installation access token (tied to a specific installation)</li>
<li>User-to-server, where the App acts on behalf of a user and authenticates using an OAuth token (tied to a combination of the user <em>and</em> any installations on accounts the user has access to)</li>
</ol>
<aside class="quote no-group" data-username="unformatt" data-post="1" data-topic="13990">
<div class="title">
<div class="quote-controls"></div>
 unformatt:</div>
<blockquote>
<p>What is “User authorization callback URL” for?</p>
</blockquote>
</aside>
<p>The “User authorization callback URL” is the URL that people are redirected to after they <a href="https://docs.github.com/en/developers/apps/identifying-and-authorizing-users-for-github-apps" rel="nofollow noopener">authorize your GitHub App to act on their behalf</a>. This is the user-to-server OAuth flow for GitHub Apps.</p>
<p>For example, a user could visit <a href="https://github.com/login/oauth/authorize?client_id=Iv1.YourAppsClientIdHere" rel="nofollow noopener">https://github.com/login/oauth/authorize?client_id=Iv1.YourAppsClientIdHere</a> and, if they agreed, click the “Authorize” button to grant your App authority to act on their behalf.</p>
<p>The “Setup URL” is the URL that people are redirected to after they <em>install</em> your App (either on their own user account, or on an organization to which they have access). Having your App installed somewhere is a prerequisite for <a href="https://docs.github.com/en/developers/apps/authenticating-with-github-apps#authenticating-as-an-installation" rel="nofollow noopener">authenticating as an installation</a> (server-to-server).</p>
<p>These two things can happen independently, so there are two different URLs in the App’s settings that you can configure.</p>
<p>If you check the “Request user authorization (OAuth) during installation” box, user’s will <em>always</em> be asked to authorize (OAuth) your App when it’s installed and they will always be redirected to your App’s “User authorization callback URL” and <em>not</em> your App’s “Setup URL”.</p>
<aside class="quote no-group" data-username="unformatt" data-post="6" data-topic="13990">
<div class="title">
<div class="quote-controls"></div>
 unformatt:</div>
<blockquote>
<p>I can’t remember how I got around it. I think when the user first installs the apps, it works (i.e. redirects) but after that, if they just updated settings, it does not redirect. I never found a way around it. Very bad UX for my users.</p>
</blockquote>
</aside>
<p>In this case, if you always want your users to go through the OAuth flow after installation or updating their installation’s settings you should:</p>
<ol>
<li>Enter a URL in the “User authorization callback URL”</li>
<li>Check the “Request user authorization (OAuth) during installation” box</li>
<li>The “Setup URL” field will be disabled at this point</li>
<li>Check the “Redirect on update” box</li>
</ol>
<p>If you don’t need to use the OAuth flow with your GitHub App, then do the following:</p>
<ol>
<li>Leave the “User authorization callback URL” field blank</li>
<li>Uncheck the “Request user authorization (OAuth) during installation” box</li>
<li>Enter a URL in the “Setup URL” field</li>
<li>Check the “Redirect on update” box</li>
</ol>
<p>Users will always be redirected to your “Setup URL” whenever your App is installed, or an installation’s settings updated (E.g. repositories added/removed).</p>

**Comment by jeevanragula** [comment]
<p><a class="mention" href="/jamesmartin">@jamesmartin</a> I have a usecase where I want users to be redirected to my site which is deployed in multiple regions based on some parameter.<br>
Eg: <a href="https://us.mysite.com" rel="noopener nofollow ugc">https://us.mysite.com</a> or <a href="https://eu.mysite.com" rel="noopener nofollow ugc">https://eu.mysite.com</a></p>
<p>How do I achieve that?<br>
Also, Can I configure multiple webhook URLs for an app?</p>

**Comment by jamesmartin** [comment]
<aside class="quote no-group" data-username="jeevanragula" data-post="9" data-topic="13990">
<div class="title">
<div class="quote-controls"></div>
 jeevanragula:</div>
<blockquote>
<p>I have a usecase where I want users to be redirected to my site which is deployed in multiple regions based on some parameter.<br>
Eg: <a href="https://us.mysite.com" rel="noopener nofollow ugc">https://us.mysite.com </a> or <a href="https://eu.mysite.com" rel="noopener nofollow ugc">https://eu.mysite.com</a></p>
<p>How do I achieve that?</p>
</blockquote>
</aside>
<p><a class="mention" href="/jeevanragula">@jeevanragula</a> there’s no simple way to achieve this currently. If you’re initiating a new installation of your App via <code>github.com/apps/{app}/installations/new/permissions</code> you can provide the <code>state</code> query parameter, which will be preserved through the installation process and appended to the query parameters of either your User authorization callback URL <em>or</em> your GitHub App Setup URL (whichever you’re using). You could use this value to determine the correct region for a subsequent redirection.</p>
<aside class="quote no-group" data-username="jeevanragula" data-post="9" data-topic="13990">
<div class="title">
<div class="quote-controls"></div>
 jeevanragula:</div>
<blockquote>
<p>Also, Can I configure multiple webhook URLs for an app?</p>
</blockquote>
</aside>
<p>No, that’s not possible today. Can you explain the need for multiple Webhook URLs, please? :thinking:</p>

**Comment by jeevanragula** [comment]
<aside class="quote no-group quote-modified" data-username="jamesmartin" data-post="10" data-topic="13990">
<div class="title">
<div class="quote-controls"></div>
 jamesmartin:</div>
<blockquote>
<aside class="quote no-group" data-username="jeevanragula" data-post="9" data-topic="13990">
<div class="title">
<div class="quote-controls"></div>
 jeevanragula:</div>
<blockquote>
<p>Also, Can I configure multiple webhook URLs for an app?</p>
</blockquote>
</aside>
</blockquote>
</aside>
<p><a class="mention" href="/jamesmartin">@jamesmartin</a> Thanks for your answer.<br>
The reason for having multiple webhooks is, We want to receive installation removal or repository change events to different endpoints depending on the region. We don’t want to create multiple apps in marketplace for each region.</p>

**Comment by jamesmartin** [comment]
<aside class="quote no-group" data-username="jeevanragula" data-post="11" data-topic="13990">
<div class="title">
<div class="quote-controls"></div>
 jeevanragula:</div>
<blockquote>
<p>We want to receive installation removal or repository change events to different endpoints depending on the region. We don’t want to create multiple apps in marketplace for each region.</p>
</blockquote>
</aside>
<p><a class="mention" href="/jeevanragula">@jeevanragula</a> thanks for explaining. How do you decide which region a GitHub App installation will be associated with? Is this decision made by the installing user? Can an installation only be associated with a single region?</p>

**Comment by zm699780** [comment]
Wow
